builtin/v10/migration/top.go

```diff
@@ -0,0 +1,280 @@
+
+	// Monitor the job queue. This non-critical goroutine is outside the errgroup and exits when
+	// workersFinished is closed, or the context done.
+	workersFinished := make(chan struct{}) // Closed when waitgroup is emptied.
+	if cfg.ProgressLogPeriod > 0 {
+		go func() {
+			defer log.Log(rt.DEBUG, "Job queue monitor done")
+			for {
+				select {
+				case <-time.After(cfg.ProgressLogPeriod):
+					jobsNow := jobCount // Snapshot values to avoid incorrect-looking arithmetic if they change.
+					doneNow := doneCount
+					pendingNow := jobsNow - doneNow
+					elapsed := time.Since(startTime)
+					rate := float64(doneNow) / elapsed.Seconds()
+					log.Log(rt.INFO, "%d jobs created, %d done, %d pending after %v (%.0f/s)",
+						jobsNow, doneNow, pendingNow, elapsed, rate)
+				case <-workersFinished:
+					return
+				case <-ctx.Done():
+					return
+				}
+			}
+		}()
+	}
+
+	// Close result channel when workers are done sending to it.
+	grp.Go(func() error {
+		workerWg.Wait()
+		close(jobResultCh)
+		close(workersFinished)
+		log.Log(rt.INFO, "All workers done after %v", time.Since(startTime))
+		return nil
+	})
+
+	// Insert migrated records in output state tree and accumulators.
+	grp.Go(func() error {
+		log.Log(rt.INFO, "Result writer started")
+		resultCount := 0
+		for result := range jobResultCh {
+			if err := actorsOut.SetActorV5(result.Address, &result.ActorV5); err != nil {
+				return err
+			}
+			resultCount++
+		}
+		log.Log(rt.INFO, "Result writer wrote %d results to state tree after %v", resultCount, time.Since(startTime))
+		return nil
+	})
+
+	if err := grp.Wait(); err != nil {
+		return cid.Undef, err
+	}
+
+	elapsed := time.Since(startTime)
+	rate := float64(doneCount) / elapsed.Seconds()
+	log.Log(rt.INFO, "All %d done after %v (%.0f/s), flushing state root.", doneCount, elapsed, rate)
+
+	return actorsOut.Flush()
+}
+
+type migrationJob struct {
+	address.Address
+	builtin.ActorV4
+	migration.ActorMigration
+	cache migration.MigrationCache
+}
+
+type migrationJobResult struct {
+	address.Address
+	builtin.ActorV5
+}
+
+func (job *migrationJob) run(ctx context.Context, store cbor.IpldStore, priorEpoch abi.ChainEpoch) (*migrationJobResult, error) {
+	result, err := job.MigrateState(ctx, store, migration.ActorMigrationInput{
+		Address:    job.Address,
+		Head:       job.ActorV4.Head,
+		PriorEpoch: priorEpoch,
+		Cache:      job.cache,
+	})
+	if err != nil {
+		return nil, xerrors.Errorf("state migration failed for actor code %s, addr %s: %w",
+			job.ActorV4.Code, job.Address, err)
+	}
+
+	// Set up new actor record with the migrated state.
+	return &migrationJobResult{
+		job.Address, // Unchanged
+		builtin.ActorV5{
+			Code:       result.NewCodeCID,
+			Head:       result.NewHead,
+			CallSeqNum: job.ActorV4.CallSeqNum, // Unchanged
+			Balance:    job.ActorV4.Balance,    // Unchanged
+			Address:    nil,                    // TODO input predictable address for accounts
```

Comment:
The TODO can stay, but I _think_ the current proposal is to not actually populate this for accounts.

Reply:
Yeah, I don't think we have anything to do here, but will keep TODO to check later.